### PR TITLE
rxKotlin for rxJava2

### PR DESCRIPTION
**Pull request by stepango**
Start working on rxKotlin for rxJava2. For now, it's just a raw port of rxKotlin. Any feedback highly appreciated.

### CODE REVIEW

**Comment by thomasnield** [comment]
Awesome! I'll take a look later...

**Comment by stepango** [comment]
@thomasnield thanks a lot

File: LICENSE [hunk 1]
```diff
@@ -180,7 +180,7 @@
 
       To apply the Apache License to your work, attach the following
       boilerplate notice, with the fields enclosed by brackets "[]"
-      replaced with your own identifying information. (Don't include
+      replaced with your own identifying information. (Don'value include
```
Comment (stepango) [inline LICENSE]: Keep `Don't`

File: build.gradle [hunk 2]
```diff
@@ -1,24 +1,26 @@
...
+    testCompile 'org.funktionale:funktionale:0.9'
     testCompile 'junit:junit:4.12'
-    testCompile 'org.mockito:mockito-core:1.8.5'
+    testCompile 'org.mockito:mockito-core:1.10.19'
     examplesCompile 'com.squareup.retrofit:retrofit:1.9.+'
 }
 
 task wrapper(type: Wrapper) {
-    gradleVersion = '2.2.1'
+    gradleVersion = '2.10'
```
Comment (stepango) [inline build.gradle]: Any chance for `nebula` to work with gradle 3+?

File: src/test/kotlin/rx/lang/kotlin/BasicKotlinTests.kt [hunk 3]
```diff
@@ -33,29 +37,25 @@ class BasicKotlinTests : KotlinTests() {
 
 
     @Test fun testCreate() {
-        Observable.create(OnSubscribe<kotlin.String> { subscriber ->
-            subscriber.onNext("Hello")
-            subscriber.onCompleted()
-        }).subscribe { result ->
+        Observable.create<kotlin.String> { onSubscribe ->
```
Comment (stepango) [inline src/test/kotlin/rx/lang/kotlin/BasicKotlinTests.kt]: `kotlin.String` -> `String`

File: src/test/kotlin/rx/lang/kotlin/ObservablesTest.kt [hunk 4]
```diff
@@ -1,34 +1,38 @@
...
-import org.junit.Assert.*
+import io.reactivex.Observable
+import io.reactivex.observers.TestObserver
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotNull
 import org.junit.Ignore
-import rx.Observable
-import rx.observers.TestSubscriber
 import java.util.concurrent.atomic.AtomicInteger
 import org.junit.Test as test
```
Comment (stepango) [inline src/test/kotlin/rx/lang/kotlin/ObservablesTest.kt]: keep `Test`

File: build.gradle (hunk 2 above)
Comment (JakeWharton) [reply build.gradle]: In RxAndroid we dropped nebula. It hurts more than it helps.

File: src/examples/kotlin/rx/lang/kotlin/examples/examples.kt [hunk 5]
```diff
@@ -1,115 +1,119 @@
...
-
+    
     asyncWiki("Tiger", "Elephant").subscribe(printArticle)
-
+    
     asyncWikiWithErrorHandling("Tiger", "Elephant").subscribe(printArticle) { e ->
         println("--- Error ---\n${e.message}")
     }
-
+
```
Comment (JakeWharton) [inline src/examples/kotlin/rx/lang/kotlin/examples/examples.kt]: What's with all the added trailing whitespace?

File: src/examples/kotlin/rx/lang/kotlin/examples/examples.kt [hunk 6]
```diff
@@ -1,115 +1,119 @@
...
-    observable { subscriber ->
-        thread {
-            articleNames.toObservable()
-                    .flatMap { name -> URL("http://en.wikipedia.org/wiki/$name").toScannerObservable().first() }
-                    .subscribe(subscriber)
-        }
+fun asyncWiki(vararg articleNames: String): Observable<String> = observable { subscriber ->
+    thread {
+        articleNames.toObservable()
+                .flatMapMaybe { name: String -> URL("http://en.wikipedia.org/wiki/$name").toScannerObservable().firstElement() }
```
Comment (JakeWharton) [inline src/examples/kotlin/rx/lang/kotlin/examples/examples.kt]: Does the argument need an explicit type?

File: src/main/kotlin/rx/lang/kotlin/observables.kt [hunk 7]
```diff
@@ -1,107 +1,98 @@
...
-fun DoubleArray.toObservable() : Observable<Double> = this.toList().toObservable()
-fun <T> Array<out T>.toObservable() : Observable<T> = Observable.from(this)
-
-fun IntProgression.toObservable() : Observable<Int> =
-        if (step == 1 && last.toLong() - first < Integer.MAX_VALUE) Observable.range(first, Math.max(0, last - first + 1))
-        else Observable.from(this)
+/**
+ * [Observable.defer] alias
+ */
+fun <T : Any> Observable<T>.defer() = Observable.defer { this }
```
Comment (JakeWharton) [inline src/main/kotlin/rx/lang/kotlin/observables.kt]: This shouldn't exist. The whole point of defer is the function is deferred to create an Observable. If you already have one there's no point in calling defer.

File: src/main/kotlin/rx/lang/kotlin/observables.kt [hunk 8]
```diff
@@ -1,107 +1,98 @@
...
-fun <T> Observable<T>.onError(block : (Throwable) -> Unit) : Observable<T> = doOnError(block)
-@Suppress("BASE_WITH_NULLABLE_UPPER_BOUND") fun <T> Observable<T>.firstOrNull() : Observable<T?> = firstOrDefault(null)
-fun <T> BlockingObservable<T>.firstOrNull() : T = firstOrDefault(null)
+fun <T : Any> Iterator<T>.toObservable(): Observable<T> = toIterable().toObservable()
+fun <T : Any> Iterable<T>.toObservable(): Observable<T> = Observable.fromIterable(this)
+fun <T : Any> Sequence<T>.toObservable(): Observable<T> = Observable.fromIterable(iterator().toIterable())
 
-@Suppress("BASE_WITH_NULLABLE_UPPER_BOUND") fun <T> Observable<T>.onErrorReturnNull() : Observable<T?> = onErrorReturn<T> {null}
+fun <T : Any> T.toSingletonObservable(): Observable<T> = Observable.just(this)
+fun <T : Any> Throwable.toObservable(): Observable<T> = Observable.error(this)
```
Comment (JakeWharton) [inline src/main/kotlin/rx/lang/kotlin/observables.kt]: These two are a bit ridiculous

File: src/examples/kotlin/rx/lang/kotlin/examples/examples.kt (hunk 5 above)
Comment (stepango) [reply src/examples/kotlin/rx/lang/kotlin/examples/examples.kt]: Looks like default kotlin formatting settings in one of the last plugin releases.

File: src/examples/kotlin/rx/lang/kotlin/examples/examples.kt (hunk 6 above)
Comment (stepango) [reply src/examples/kotlin/rx/lang/kotlin/examples/examples.kt]: Nope

File: src/main/kotlin/rx/lang/kotlin/observables.kt (hunk 7 above)
Comment (stepango) [reply src/main/kotlin/rx/lang/kotlin/observables.kt]: Agreed, forgot to delete it before commit

File: src/main/kotlin/rx/lang/kotlin/observables.kt (hunk 8 above)
Comment (stepango) [reply src/main/kotlin/rx/lang/kotlin/observables.kt]: yes

File: src/main/kotlin/rx/lang/kotlin/subscribers.kt [hunk 9]
```diff
@@ -1,19 +1,21 @@
...
-class FunctionSubscriber<T>() : Subscriber<T>() {
-    private val onCompletedFunctions = ArrayList<() -> Unit>()
+import io.reactivex.CompletableObserver
+import io.reactivex.MaybeObserver
+import io.reactivex.Observer
+import io.reactivex.SingleObserver
+import io.reactivex.disposables.Disposable
+import java.util.ArrayList
+
+class FunctionSubscriber<T : Any> : Observer<T>, MaybeObserver<T>, SingleObserver<T>, CompletableObserver {
```
Comment (stepango) [inline src/main/kotlin/rx/lang/kotlin/subscribers.kt]: Reimplement using `OnSubscribe*` interfaces

File: src/examples/kotlin/rx/lang/kotlin/examples/examples.kt [hunk 10]
```diff
@@ -43,73 +50,70 @@ fun main(args: Array<String>) {
     addToCompositeSubscription()
 }
 
-private fun URL.toScannerObservable() = observable<String> { s ->
+private fun URL.toScannerObservable() = observable<String>({ s ->
```
Comment (stepango) [inline src/examples/kotlin/rx/lang/kotlin/examples/examples.kt]: remove ()

**Comment by stepango** [comment]
Porting `subscribeWith` was a good exercise but looks like i'll drop it from rxKotlin 2.x and rxKorlin 1.x as well. Probably release it as a separate artefact later.

**Comment by thomasnield** [comment]
I was thinking about that for a few days. Makes sense but it's too bad. Is there a way to present each action as a named parameter to `subscribe()`?

**Comment by thomasnield** [comment]
```kotlin
import io.reactivex.Observable

fun main(args: Array<String>) {

    Observable.just("Alpha", "Beta", "Gamma")
            .subscribe(onNext = ::println, onComplete = { println("done") })

}

fun <T> Observable<T>.subscribe(onNext: ((T) -> Unit)? = null,
                                onError: ((Throwable) -> Unit)? = null,
                                onComplete: (() -> Unit)? = null) = subscribe(onNext, onError, onComplete)


```

**Comment by stepango** [comment]
@thomasnield that's exactly what I'm thinking about. The only problem is - the name of a function, ide will not auto-import it because of the name clash. `subscribeWith` is taken already and has different semantics in rxJava2. As an option, function could be named `subscribeBy`. Any ideas?

**Comment by stepango** [comment]
RxJava2 don't accept null params in `subscribe()`

**Comment by thomasnield** [comment]
RxJava2 doesn't accept null anything it seems. I guess you can just create a hard observer in the implementation then...

**Comment by iNoles** [comment]
@stepango According to https://kotlinlang.org/docs/reference/packages.html, you can use alias for name clash such as like "import bar.Bar as bBar // bBar stands for 'bar.Bar'"

**Comment by stepango** [comment]
@iNoles I know about import aliases, do you have some particular use case in mind?

**Comment by iNoles** [comment]
@stepango I was thinking about to use rxjava as import alias.

**Comment by stepango** [comment]
@iNoles don't get it. What exactly do you mean? Any code examples?

**Comment by thomasnield** [comment]
@stepango is it possible you can merge what you have into 2.x? I'd like to start contributing to this and it might be easier to centralize it into the repository.

**Comment by stepango** [comment]
@thomasnield was just about to do that. Finally, have some free time this weekend.

**Pull request merged**